token/src/components/token/erc721/erc721_balance.cairo

```diff
@@ -0,0 +1,228 @@
+        }
+
+        fn transferFrom(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            tokenId: u128
+        ) {
+            self.transfer_from(from, to, tokenId)
+        }
+        fn safeTransferFrom(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            tokenId: u128,
+            data: Span<felt252>
+        ) {
+            self.safe_transfer_from(from, to, tokenId, data)
+        }
+    }
+
+
+    #[generate_trait]
+    impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        +IWorldProvider<TContractState>,
+        impl ERC721Approval: erc721_approval_comp::HasComponent<TContractState>,
+        impl ERC721Owner: erc721_owner_comp::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        fn get_balance(
+            self: @ComponentState<TContractState>, account: ContractAddress
+        ) -> ERC721BalanceModel {
+            get!(
+                self.get_contract().world(), (get_contract_address(), account), (ERC721BalanceModel)
+            )
+        }
+
+        fn set_balance(
+            self: @ComponentState<TContractState>, account: ContractAddress, amount: u256
+        ) {
+            set!(
+                self.get_contract().world(),
+                ERC721BalanceModel { token: get_contract_address(), account, amount }
+            );
+        }
+
+        fn transfer_internal(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u128
+        ) {
+            assert(!to.is_zero(), Errors::INVALID_RECEIVER);
+            let mut erc721_approval = get_dep_component_mut!(ref self, ERC721Approval);
+            let mut erc721_owner = get_dep_component_mut!(ref self, ERC721Owner);
+
+            let owner = erc721_owner.get_owner(token_id).address;
+            assert(from == owner, Errors::WRONG_SENDER);
+
+            // Implicit clear approvals, no need to emit an event
+            erc721_approval.set_token_approval(owner, Zeroable::zero(), token_id, false);
+
+            self.set_balance(from, self.get_balance(from).amount - 1);
+            self.set_balance(to, self.get_balance(to).amount + 1);
+            erc721_owner.set_owner(token_id, to);
+
+            let transfer_event = Transfer { from, to, token_id };
+
+            self.emit(transfer_event.clone());
+            emit!(self.get_contract().world(), (Event::Transfer(transfer_event)));
+        }
+
+        fn safe_transfer_internal(
+            ref self: ComponentState<TContractState>,
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u128,
+            data: Span<felt252>
+        ) {
+            self.transfer_internal(from, to, token_id);
```

Comment:
i don't remind the exact discussion about why we didn't support the 'safe' part in prev impl
ideally it should check that receiver implements IERC721Receiver